admin/services/service-manager/attendance/service-manager.attendance.service.ts

```diff
@@ -0,0 +1,23 @@
+import { AttendanceCode, ServiceManagerAttendanceDataService } from './service-manager.attendance.data.service'
+
+export class ServiceManagerAttendanceService {
+  static async getAttendanceCodes (): Promise<AttendanceCode[]> {
+    return ServiceManagerAttendanceDataService.getAttendanceCodes()
+  }
+
+  /**
+   * @description Sets the visibility of the attendance codes to true if they are included in the set.
+   * All other non privileged attendance codes will be set to not visible.
+   * @param visibleAttendanceCodes - The attendance codes to be visible
+   */
+  static async setVisibleAttendanceCodes (visibleAttendanceCodes: string[] | undefined): Promise<void> {
+    if (!visibleAttendanceCodes || visibleAttendanceCodes.length === 0) {
+      visibleAttendanceCodes = []
+    }
+    const allAttendanceCodes = await ServiceManagerAttendanceDataService.getAttendanceCodes()
+    for (const attendanceCode of allAttendanceCodes) {
+      const visible = visibleAttendanceCodes.includes(attendanceCode.code)
+      await ServiceManagerAttendanceDataService.setVisibility(attendanceCode.code, visible)
```

Comment:
I would normally discourage making sql calls in a for loop.  This should really be one call to set all attendance codes in one go, or perhaps all modified attendance codes. However, as the service manager will only do this very rarely it will probably be okay. 🤔

Reply:
yes, i brooded on this for a while, and its extremely unlikely that there will ever be enough entries to cause a concerning amount of calls.  Plus, as you say - its for the service manager and unlikely to be used once a year, if at all.

Reply:
i decided to refactor it into one transaction for all the updates.  i'd rather set the right example!